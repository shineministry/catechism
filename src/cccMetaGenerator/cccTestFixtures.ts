import { CCCStore } from '../store/cccTypedefs'
import {
  createTocLink,
  createTocNode,
  createPageNode,
  createParagraphWithCCCRefElements,
  createParagraph,
} from './testHelpers'

export const createMockCCC = (): CCCStore =>
  (({
    toc_link_tree: [
      createTocLink('toc-1', [
        createTocLink('toc-2'),
        createTocLink('toc-3'),
        createTocLink('toc-no-page-1'),
      ]),
      createTocLink('toc-10'),
      createTocLink('toc-no-page-2'),
    ],
    toc_nodes: {
      'toc-1': createTocNode('toc-1', 'link text 1'),
      'toc-2': createTocNode('toc-2', 'link text 2'),
      'toc-3': createTocNode('toc-3', 'link text 3'),
      'toc-no-page-1': createTocNode('toc-no-page-1', 'link text x'),
      'toc-10': createTocNode('toc-10', 'link text 10'),
      'toc-no-page-2': createTocNode('toc-no-page-2', 'link text x'),
    },
    page_nodes: {
      'toc-1': createPageNode('toc-1', [
        createParagraphWithCCCRefElements([1, 3]),
      ]),
      'toc-2': createPageNode('toc-2', [createParagraph()]),
      'toc-3': createPageNode('toc-3', [
        createParagraphWithCCCRefElements([4, 6]),
      ]),
      'toc-10': createPageNode('toc-10', [
        createParagraphWithCCCRefElements([7, 8]),
      ]),
    },
  } as unknown) as CCCStore)
