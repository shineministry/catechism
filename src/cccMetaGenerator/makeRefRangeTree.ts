import { getAllTocLinks } from './cccMetaGenerator'
import {
  PageNode,
  CCCRefElement,
  TOCLink,
  PageNodes,
} from './../store/cccTypedefs'
export interface CCCRefRangeTree {
  root: RangeTreeNode
}
export interface RangeTreeNode {
  left: RangeTreeNode | undefined
  right: RangeTreeNode | undefined
  min: number
  max: number
  tocId: string | undefined
}

export const makeCCCRefRangeTree = (
  tocLinkTree: TOCLink[],
  pageNodes: PageNodes
): CCCRefRangeTree => {
  const tocIds = getAllTocLinks(tocLinkTree)
  const tocIdsWithPages = tocIds.filter(tocId => tocId in pageNodes)

  const rangeNodes: RangeTreeNode[] = tocIdsWithPages
    .map(tocId => pageNodes[tocId])
    .map(getRefRangeForPage)
    .filter(n => !!n) as RangeTreeNode[]

  /* eslint-disable  */
  let remainingNodes: RangeTreeNode[] = [...rangeNodes]
  let tempNodes: RangeTreeNode[] = []

  while (remainingNodes.length > 1) {
    const evenIndexes = remainingNodes.map((_, i) => i).filter(i => i % 2 === 0)
    evenIndexes.forEach(i => {
      const currentNode = remainingNodes[i]

      const hasNextElement = !!remainingNodes[i + 1]
      if (!hasNextElement) {
        tempNodes.push(currentNode)
        return
      }

      const nextNode = remainingNodes[i + 1]

      const newParentNode: RangeTreeNode = {
        left: currentNode,
        right: nextNode,
        min: currentNode.min,
        max: nextNode.max,
        tocId: undefined,
      }
      tempNodes.push(newParentNode)
    })

    remainingNodes = [...tempNodes]
    tempNodes = []
  }

  /* eslint-enable  */
  const rootNode = remainingNodes[0]
  return { root: rootNode }
}

const getRefRangeForPage = (pageNode: PageNode): RangeTreeNode | undefined => {
  const { id, paragraphs } = pageNode

  if (paragraphs.length === 0) {
    return undefined
  }

  const paragraphElements = paragraphs
    .map(p => p.elements)
    .reduce((acc, curr) => [...acc, ...curr], [])
  const cccRefNumbers = paragraphElements
    .filter(e => e.type === 'ref-ccc')
    .map(e => {
      return (e as CCCRefElement).ref_number
    })
    .filter(n => n)

  if (cccRefNumbers.length === 0) {
    return undefined
  }

  return {
    left: undefined,
    right: undefined,
    min: Math.min.apply(null, cccRefNumbers),
    max: Math.max.apply(null, cccRefNumbers),
    tocId: id,
  }
}

export const findPage = (
  cccRefNumber: number,
  rootRangeNode: RangeTreeNode
): string | undefined => {
  /* eslint-disable  */
  let currentNode: RangeTreeNode = rootRangeNode

  while (true) {
    const { left, right } = currentNode
    const isInLeft = !!left ? cccRefNumber <= left.max : false
    const isInRight = !!right ? cccRefNumber >= right.min : false

    if (isInLeft) {
      currentNode = left as RangeTreeNode
    } else if (isInRight) {
      currentNode = right as RangeTreeNode
    } else {
      return undefined
    }

    if (!!currentNode.tocId) {
      break
    }
  }

  /* eslint-enable  */

  return currentNode.tocId
}
