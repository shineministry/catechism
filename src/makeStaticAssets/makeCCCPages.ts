import {
  filterEmptyAttrs,
  filterTrailingEmptyParagraphs,
} from './makeCCCPagesUtils'
import {
  PageNode,
  PageParagraphElement,
  PageParagraph,
  TextElement,
  AnchorElement,
} from './../store/cccTypedefs'
import { stripUrlShortLink } from './../cccMetaGenerator/makeUrlMap'
import { CCCStore } from '../store/cccTypedefs'
import {
  CCCMeta,
  LeanPageNode,
  LeanPageParagraph,
  LeanPageParagraphElement,
} from './typedefs'

export const makeCCCPages = async (
  ccc: CCCStore,
  cccMeta: CCCMeta
): Promise<CCCExportedPage[]> => {
  const { page_nodes } = ccc
  const { urlMap } = cccMeta

  const filterNullPages = (fullUrl: string) => {
    const tocId = urlMap[fullUrl]
    const page = page_nodes[tocId]
    return !!page
  }

  return Object.keys(urlMap)
    .filter(filterNullPages)
    .map(fullUrl => {
      const shortUrl = stripUrlShortLink(fullUrl)
      const tocId = urlMap[fullUrl]
      const page = page_nodes[tocId]

      const leanPage = makeLeanPage(page)
      return { fileName: shortUrl, jsonContent: JSON.stringify(leanPage) }
    })
    .filter(page => !!page.jsonContent)
}

const makeLeanPage = (page: PageNode): LeanPageNode => {
  const { paragraphs } = page

  const withoutEmptyTrailingParagraphs = filterTrailingEmptyParagraphs(
    paragraphs
  )
  const leanParagraphs = withoutEmptyTrailingParagraphs.map(makeLeanParagraph)
  return { ...page, paragraphs: leanParagraphs }
}

const makeLeanParagraph = (paragraph: PageParagraph): LeanPageParagraph => {
  const { attrs, elements } = paragraph
  const leanElements = elements.map(makeLeanParagraphElement)

  return { elements: leanElements, ...filterEmptyAttrs(attrs) }
}

const makeLeanParagraphElement = (
  element: PageParagraphElement
): LeanPageParagraphElement => {
  const { type } = element

  if (type === 'text') {
    const { attrs, ...rest } = element as TextElement
    return { ...filterEmptyAttrs(attrs), ...rest }
  }

  if (type === 'ref-anchor') {
    const { attrs, ...rest } = element as AnchorElement
    return { ...filterEmptyAttrs(attrs), ...rest }
  }

  return element
}

export interface CCCExportedPage {
  fileName: string
  jsonContent: string
}
