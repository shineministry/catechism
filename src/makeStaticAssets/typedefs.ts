import { TextElement, AnchorElement } from './../store/cccTypedefs'
import { LeanPageParagraph } from './typedefs'
import {
  CCCStore,
  PageElementAttributes,
  RefElement,
  CCCRefElement,
  SpacerElement,
  TOCLink,
  PageNode,
  PageParagraph,
} from '../store/cccTypedefs'
import { PageMetaMap } from '../cccMetaGenerator/makePageMetaMap'
import { UrlToTocIdMap } from '../cccMetaGenerator/makeUrlMap'
import { CCCRefRangeTree } from '../cccMetaGenerator/makeRefRangeTree'
import { BreadcrumbsMap } from '../cccMetaGenerator/makeBreadcrumbs'

export interface CCCEnhancedStore {
  store: CCCLeanStore
  extraMeta: CCCMeta
}

export type CCCLeanStore = Omit<CCCStore, 'page_nodes' | 'toc_link_tree'> & {
  toc_link_tree: LeanTOCLink[]
}

export type LeanTOCLink = Omit<TOCLink, 'children'> & {
  children?: LeanTOCLink[]
}

export type LeanPageNode = Omit<PageNode, 'paragraphs'> & {
  paragraphs: LeanPageParagraph[]
}

export type LeanPageParagraph = Omit<PageParagraph, 'elements' | 'attrs'> & {
  elements: LeanPageParagraphElement[]
  attrs?: PageElementAttributes
}

export type LeanPageParagraphElement =
  | LeanTextElement
  | RefElement
  | LeanAnchorElement
  | CCCRefElement
  | SpacerElement

export type LeanTextElement = Omit<TextElement, 'attrs'> & {
  attrs?: PageElementAttributes
}

export type LeanAnchorElement = Omit<AnchorElement, 'attrs'> & {
  attrs?: PageElementAttributes
}

export interface CCCMeta {
  pageMetaMap: PageMetaMap
  urlMap: UrlToTocIdMap
  cccRefRangeTree: CCCRefRangeTree
  breadcrumbsMap: BreadcrumbsMap
}

export interface TocIdToUrlMap {
  // maps tocId to url
  [tocId: string]: string
}
