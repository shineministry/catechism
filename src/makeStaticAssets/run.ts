import { makeStaticAssets } from './makeStaticAssets'

makeStaticAssets()
