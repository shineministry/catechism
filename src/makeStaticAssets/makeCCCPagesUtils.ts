import { LeanPageParagraph } from './typedefs'
import { PageElementAttributes, PageParagraph } from '../store/cccTypedefs'

export const filterTrailingEmptyParagraphs = (
  paragraphs: PageParagraph[]
): PageParagraph[] => {
  const indexesToFilter = getTrailingEmptyParagraphIndexes(paragraphs)
  return paragraphs.filter((_, index) => !indexesToFilter.has(index))
}

const getTrailingEmptyParagraphIndexes = (
  paragraphs: LeanPageParagraph[]
): Set<number> => {
  const emptyParagraphIndexes = paragraphs.reduceRight<
    TrailingEmptyParagraphAccumulator
  >(
    (acc, paragraph, index) => {
      if (acc.ignoreRest) {
        return acc
      }

      if (isEmptyParagraph(paragraph)) {
        return { ...acc, indexes: [...acc.indexes, index] }
      }

      return { ...acc, ignoreRest: true }
    },
    { ignoreRest: false, indexes: [] }
  )

  return new Set(emptyParagraphIndexes.indexes)
}

const isEmptyParagraph = (paragraph: LeanPageParagraph): boolean => {
  const { elements } = paragraph
  if (elements.length === 0) {
    return true
  }

  if (elements.length === 1) {
    const [firstElement] = elements
    if (firstElement.type === 'spacer') {
      return true
    }
  }

  return false
}

interface TrailingEmptyParagraphAccumulator {
  ignoreRest: boolean
  indexes: number[]
}

export const filterEmptyAttrs = (attrs: PageElementAttributes) => {
  const isEmptyAttrs = Object.keys(attrs).length === 0
  return isEmptyAttrs ? {} : { attrs }
}
