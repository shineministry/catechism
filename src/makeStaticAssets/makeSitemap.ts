import { createSitemap, ISitemapItemOptionsLoose } from 'sitemap'
import { CCCMeta } from './typedefs.js'

const SITE_HOMEPAGE = 'https://nossbigg.github.io/catechism'
const makeIndexUrl = () => `${SITE_HOMEPAGE}/#/index`
const makePageUrl = (pageUrl: string) => `${SITE_HOMEPAGE}/#/p/${pageUrl}`

export const makeSitemapXml = (cccMeta: CCCMeta): string => {
  const indexItem: ISitemapItemOptionsLoose = { url: makeIndexUrl() }
  const pageItems: ISitemapItemOptionsLoose[] = Object.keys(cccMeta.urlMap)
    .map(makePageUrl)
    .map(url => ({ url }))

  const sitemap = createSitemap({
    hostname: SITE_HOMEPAGE,
    urls: [indexItem, ...pageItems],
  })
  return sitemap.toXML()
}
