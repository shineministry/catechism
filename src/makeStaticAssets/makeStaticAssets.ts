import request from 'request-promise-native'
import fs from 'fs-extra'
import path from 'path'
import { execSync } from 'child_process'
import { CCCStore, TOCLink } from 'store/cccTypedefs'
import { makeCCCMeta } from '../cccMetaGenerator/cccMetaGenerator'
import { makeCCCPages, CCCExportedPage } from './makeCCCPages'
import { CCCLeanStore, CCCMeta, LeanTOCLink } from './typedefs'
import { makeSitemapXml } from './makeSitemap'

export const makeStaticAssets = async () => {
  await prepareDirectory()

  const ccc = await getCCCReleaseFromRemote()
  const cccMeta = await makeCCCMetadata(ccc)
  const cccPages = await makeCCCPages(ccc, cccMeta)
  const sitemapXml = makeSitemapXml(cccMeta)
  const leanCCC = makeLeanCCC(ccc)

  await saveRepoVersionFile(ccc)
  await saveCCC(leanCCC)
  await saveCCCMeta(cccMeta)
  await saveCCCPages(cccPages)
  await saveSitemapXml(sitemapXml)
}

const getCCCReleaseFromRemote = async (): Promise<CCCStore> => {
  log('ccc: downloading from remote...')
  const latestReleaseMeta: GithubAPIReleaseResponse = await request({
    uri:
      'https://api.github.com/repos/nossbigg/catechism-ccc-json/releases/latest',
    ...GITHUB_REQUEST_PARAMETERS,
  })

  const cccJsonAsset = latestReleaseMeta.assets.filter(
    asset => asset.name === 'ccc.json'
  )
  const cccDownloadUrl = cccJsonAsset[0].browser_download_url
  const cccDownload: CCCStore = await request({
    uri: cccDownloadUrl,
    ...GITHUB_REQUEST_PARAMETERS,
  })

  log('ccc: done!')
  return cccDownload
}

const makeLeanCCC = (ccc: CCCStore): CCCLeanStore => {
  const { page_nodes, toc_link_tree, ...rest } = ccc
  return { ...rest, toc_link_tree: toc_link_tree.map(makeLeanTOCLink) }
}

const makeLeanTOCLink = (link: TOCLink): LeanTOCLink => {
  const { children, ...rest } = link
  return { ...rest, ...makeLeanTOCLinkChildren(children) }
}

const makeLeanTOCLinkChildren = (children: TOCLink[]) => {
  if (children.length === 0) {
    return {}
  }
  return { children: children.map(makeLeanTOCLink) }
}

const makeCCCMetadata = async (ccc: CCCStore): Promise<CCCMeta> => {
  log('ccc meta: generating ccc meta...')
  const meta = makeCCCMeta(ccc)
  log('ccc meta: done!')
  return meta
}

const saveCCC = async (ccc: CCCLeanStore) => {
  log('ccc: saving to disk...')
  await fs.writeFile(STATIC_ASSETS_PATHS.ccc, JSON.stringify(ccc))
  log('ccc: done!')
}

const saveCCCMeta = async (cccMeta: CCCMeta) => {
  log('ccc meta: saving to disk...')
  await fs.writeFile(STATIC_ASSETS_PATHS.cccMeta, JSON.stringify(cccMeta))
  log('ccc meta: done!')
}

const saveCCCPages = async (cccPages: CCCExportedPage[]) => {
  log('ccc pages: saving to disk...')
  await Promise.all(
    cccPages.map(page => {
      const { fileName, jsonContent } = page
      const fullFileName = STATIC_ASSETS_PATHS.makePagePath(fileName)
      return fs.writeFile(fullFileName, jsonContent)
    })
  )
  log('ccc pages: done!')
}

const saveSitemapXml = async (sitemapXml: string): Promise<void> => {
  log('sitemap: saving to disk...')
  await fs.writeFile(STATIC_ASSETS_PATHS.sitemapXml, sitemapXml)
  log('sitemap: done!')
}

const saveRepoVersionFile = async (ccc: CCCStore): Promise<void> => {
  log('repo version: saving to disk...')

  const revision = execSync('git rev-parse --short HEAD')
    .toString()
    .trim()
  const versionInfo: RepoVersion = { repo: revision, store: ccc.meta.version }

  await fs.writeFile(
    STATIC_ASSETS_PATHS.versionFile,
    JSON.stringify(versionInfo)
  )
  log('repo version: done!')
}

const prepareDirectory = async () => {
  log('cleanup: cleaning existing assets...')
  await fs.remove(STATIC_ASSETS_PATHS.rootDir)
  await fs.mkdirp(STATIC_ASSETS_PATHS.rootDir)
  await fs.mkdirp(STATIC_ASSETS_PATHS.cccPagesDir)
  log('cleanup: done!')
}

// eslint-disable-next-line no-console
const log = (msg: string) => console.log(msg)

const STATIC_ASSETS_ROOT = './public/ccc'
const STATIC_ASSETS_PATHS = {
  rootDir: STATIC_ASSETS_ROOT,
  ccc: path.join(STATIC_ASSETS_ROOT, 'ccc.json'),
  cccMeta: path.join(STATIC_ASSETS_ROOT, 'cccMeta.json'),
  cccPagesDir: path.join(STATIC_ASSETS_ROOT, 'pages'),
  versionFile: path.join(STATIC_ASSETS_ROOT, 'version.json'),
  sitemapXml: path.join(STATIC_ASSETS_ROOT, 'sitemap.xml'),
  makePagePath: (fileName: string) =>
    path.join(STATIC_ASSETS_ROOT, 'pages', `${fileName}.json`),
}

const GITHUB_REQUEST_PARAMETERS = {
  method: 'GET',
  json: true,
  headers: {
    'User-Agent': 'nossbigg/catechism-web',
  },
}

interface GithubAPIReleaseResponse {
  assets: Asset[]
}

interface Asset {
  name: string
  browser_download_url: string
}

interface RepoVersion {
  repo: string
  store: string
}
