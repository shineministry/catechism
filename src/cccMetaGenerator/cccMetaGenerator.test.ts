import { createMockCCC } from './cccTestFixtures'
import {
  mergeObjectsProperties,
  parseToShortLinkText,
  makeCCCMeta,
  makeTocToUrlMap,
} from './cccMetaGenerator'
import { RangeTreeNode } from './makeRefRangeTree'
import { TocIdToUrlMap } from 'makeStaticAssets/typedefs'

describe('cccMetaGenerator', () => {
  describe('generating page meta map', () => {
    const doAction = () => {
      const cccStore = createMockCCC()
      const metadata = makeCCCMeta(cccStore)
      return metadata.pageMetaMap
    }
    const resultMap = doAction()

    it('maps url to meta node', () => {
      expect(resultMap['toc-1'].url).toEqual('0+link-text-1')
    })

    it('generates meta node with next/prev refs', () => {
      expect(resultMap['toc-3']).toEqual(
        expect.objectContaining({
          id: 'toc-3',
          prev: 'toc-2',
          next: 'toc-10',
        })
      )
    })

    it('generates first meta node with no prev ref', () => {
      expect(resultMap['toc-1']).toEqual(
        expect.objectContaining({
          id: 'toc-1',
          prev: '',
          next: 'toc-2',
        })
      )
    })

    it('generates last meta node with no next ref', () => {
      expect.objectContaining({
        id: 'toc-10',
        prev: 'toc-3',
        next: '',
      })
    })

    it('skips a node which does not have a page', () => {
      expect('toc-doesnt-exist-1' in resultMap).toBe(false)
      expect('toc-doesnt-exist-2' in resultMap).toBe(false)
    })
  })

  describe('generating url map', () => {
    const doAction = () => {
      const cccStore = createMockCCC()
      const metadata = makeCCCMeta(cccStore)
      return metadata.urlMap
    }
    const resultMap = doAction()

    it('generates a correct url map with truncated urls', () => {
      expect('0.1' in resultMap).toBe(true)
      expect(resultMap['0.1']).toEqual('toc-2')
    })
  })

  describe('generating breadcrumbs map', () => {
    const doAction = () => {
      const cccStore = createMockCCC()
      const metadata = makeCCCMeta(cccStore)
      return metadata.breadcrumbsMap
    }
    const breadcrumbMap = doAction()

    it('generates a breadcrumb', () => {
      const toc1 = breadcrumbMap['toc-1']
      expect(toc1.id).toEqual('toc-1')
      expect(toc1.parent).not.toBeUndefined()
    })

    it('generates a top-level breadcrumb', () => {
      const expectedKeysWithNoParent = ['toc-1', 'toc-10', 'toc-no-page-2']
      expectedKeysWithNoParent.forEach(key =>
        expect(breadcrumbMap[key].parent).toEqual('')
      )
    })

    it('generates a child breadcrumb', () => {
      const expectedChildPagesOfToc1 = ['toc-2', 'toc-3', 'toc-no-page-1']
      expectedChildPagesOfToc1.forEach(key =>
        expect(breadcrumbMap[key].parent).toEqual('toc-1')
      )
    })
  })

  describe('generating ref range tree', () => {
    const doAction = () => {
      const cccStore = createMockCCC()
      const metadata = makeCCCMeta(cccStore)
      return metadata.cccRefRangeTree.root
    }
    const rootNode = doAction()

    it('generates a top-level element', () => {
      expect(rootNode.min).toBe(1)
      expect(rootNode.max).toBe(8)
      expect(rootNode.tocId).toBeUndefined()
    })

    it('generates a leaf element', () => {
      const { right } = rootNode
      expect((right as RangeTreeNode).tocId).toBe('toc-10')
    })

    it('matches snapshot', () => {
      expect(rootNode).toMatchSnapshot()
    })
  })

  describe('generateTocToUrlMap', () => {
    const doAction = (): TocIdToUrlMap => {
      const cccStore = createMockCCC()
      return makeTocToUrlMap(cccStore.toc_link_tree, cccStore.toc_nodes)
    }
    const resultMap = doAction()

    const assertEntryPresent = (
      expectedPageMapping: string,
      expectedUrl: string
    ) => {
      expect(expectedPageMapping in resultMap).toBe(true)
      expect(resultMap[expectedPageMapping]).toEqual(expectedUrl)
    }

    it('generates a top-level node url', () => {
      assertEntryPresent('toc-1', '0+link-text-1')
      assertEntryPresent('toc-10', '1+link-text-10')
    })

    it('generates a nested level node url', () => {
      assertEntryPresent('toc-2', '0.1+link-text-2')
      assertEntryPresent('toc-3', '0.2+link-text-3')
    })

    it('skips a node which does not have a page', () => {
      expect('toc-doesnt-exist-1' in resultMap).toBe(false)
      expect('toc-doesnt-exist-2' in resultMap).toBe(false)
    })
  })

  describe('parseToShortLinkText', () => {
    it('parses the following titles correctly', () => {
      expect(parseToShortLinkText('SECTION TWO I. THE CREEDS')).toEqual(
        'the-creeds'
      )
      expect(
        parseToShortLinkText("CHAPTER ONE MAN'S CAPACITY FOR GOD")
      ).toEqual('mans-capacity-for-god')
      expect(
        parseToShortLinkText('Article 5 THE MORALITY OF THE PASSIONS')
      ).toEqual('the-morality-of-the-passions')
      expect(parseToShortLinkText('Paragraph 2. THE FATHER')).toEqual(
        'the-father'
      )
      expect(parseToShortLinkText('PART ONE: THE PROFESSION OF FAITH')).toEqual(
        'the-profession-of-faith'
      )
    })

    it('parses only up to the correct number of words in a link title', () => {
      expect(
        parseToShortLinkText(
          'Paragraph 2. "CONCEIVED BY THE POWER OF THE HOLY SPIRIT AND BORN OF THE VIRGIN MARY"'
        )
      ).toEqual('conceived-by-the-power-of')
    })
  })

  describe('mergeObjectsProperties', () => {
    it('merges object properties', () => {
      expect(mergeObjectsProperties([{ a: 1 }, { b: 1 }])).toEqual({
        a: 1,
        b: 1,
      })
    })
  })
})
