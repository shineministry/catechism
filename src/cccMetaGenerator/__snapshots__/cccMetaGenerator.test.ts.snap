// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cccMetaGenerator generating ref range tree matches snapshot 1`] = `
Object {
  "left": Object {
    "left": Object {
      "left": undefined,
      "max": 3,
      "min": 1,
      "right": undefined,
      "tocId": "toc-1",
    },
    "max": 6,
    "min": 1,
    "right": Object {
      "left": undefined,
      "max": 6,
      "min": 4,
      "right": undefined,
      "tocId": "toc-3",
    },
    "tocId": undefined,
  },
  "max": 8,
  "min": 1,
  "right": Object {
    "left": undefined,
    "max": 8,
    "min": 7,
    "right": undefined,
    "tocId": "toc-10",
  },
  "tocId": undefined,
}
`;
